native/cocos/renderer/gfx-base/GFXDef.cpp

```diff
@@ -35,52 +35,72 @@ namespace gfx {
 
 // T must have no implicit padding
 template <typename T>
-size_t quickHashTrivialStruct(const T *info, size_t count = 1) {
+ccstd::hash_t quickHashTrivialStruct(const T *info, size_t count = 1) {
     static_assert(std::is_trivially_copyable<T>::value && sizeof(T) % 8 == 0, "T must be 8 bytes aligned and trivially copyable");
-    return boost::hash_range(reinterpret_cast<const uint64_t *>(info), reinterpret_cast<const uint64_t *>(info + count));
+    return ccstd::hash_range(reinterpret_cast<const uint64_t *>(info), reinterpret_cast<const uint64_t *>(info + count));
 }
 
 template <>
-size_t Hasher<ColorAttachment>::operator()(const ColorAttachment &info) const {
+ccstd::hash_t Hasher<ColorAttachment>::hashValue(const ColorAttachment &info) {
     return quickHashTrivialStruct(&info);
 }
 
+template <>
+size_t Hasher<ColorAttachment>::operator()(const ColorAttachment &info) const {
+    return static_cast<size_t>(hashValue(info));
+}
+
 bool operator==(const ColorAttachment &lhs, const ColorAttachment &rhs) {
     return !memcmp(&lhs, &rhs, sizeof(ColorAttachment));
 }
 
 template <>
-size_t Hasher<DepthStencilAttachment>::operator()(const DepthStencilAttachment &info) const {
+ccstd::hash_t Hasher<DepthStencilAttachment>::hashValue(const DepthStencilAttachment &info) {
```

Comment:
Why need to add this function?

Reply:
The reason is the same as what I have replied.
> I think it's probably not, the operator() is used for interacting with `stl containers`.

```c++
ccstd::unordered_map<Vec4, gfx::InputAssembler *, Hasher<Vec4>> _quadIA;
```

So I defined two functions, one is functor for interacting with `stl containers` which returns size_t.
Another is a static function for used in our internal code.
```c++
template <typename T, typename Enable = std::enable_if_t<std::is_class<T>::value>>
struct Hasher final {
    size_t operator()(const T &info) const;
    static ccstd::hash_t hashValue(const T &info);
};
```

Using static `hashValue` function could also avoid constructing a temporary `Hasher` object.


hashValue returns ccstd::hash_t and is used in our engine code.
`size_t operator()` is used for custom hash of std containers since std containers use size_t for hash. If using ccstd::hash_t, it may cause warnings.

Reply:
Furthermore, keeping `size_t operator()` will make it easier to convert 64 bits hash to 32 bits.
Suppose we wanna use 64 bits hash on 32 bits devices some day. Then, std containers will still use `size_t` as hash value which is 32 bits on 32 bits architecture. We could do the conversion in operator() method like this:

```c++
template <>
ccstd::hash_t Hasher<SubpassDependency>::hashValue(const SubpassDependency &info) {
    return quickHashTrivialStruct(&info);
}

template <>
size_t Hasher<SubpassDependency>::operator()(const SubpassDependency &info) const {
    ccstd::hash_t value = hashValue(info); // ccstd::hash_t  is 64bits, size_t is 32bits
    return static_cast<size_t>((value >> 32) ^ (value & 0xFFFFFFFF) ); // Convert 64 bits hash to 32 bits
}
```

Reply:
I think just return hash_t for operator makes things simple. hast_t is a special type used in engine, then the corresponding Hasher is a special type too. 

All codes use the same result is better. If using two functions, then some codes will use `operator()`, some use `hashValue`, which makes things deprecated. We let JS and C++ use the same type, then use hash_t, we should also use hash_t for all codes too.

Reply:
It's the best situation we hope. But remember that, stl container ( std::unordered_map ) is external code that will use `operator()`, it's out of our control. 

If we use hash_t for the return value of `operator()`, 
- 1. it will generate compiler warning of type conversion
- 2. If we use 64 bits hash someday, hash value may be cut from 64 bits to 32 bits in std container code without notify us, it will be bug prone.

Reply:
> it will generate compiler warning of type conversion

I think it is ok, because it doesn't loose any information. Does it block the CI?

> If we use 64 bits hash someday, hash value may be cut from 64 bits to 32 bits in std container code without notifying us, it will be bug prone.

We can add some comments. And in current implementation, we also need to remove the `hashValue` function too.

Reply:
DONE in the latest commit

Reply:
DONE